### DISCUSSION TITLE
Using input data from the United Nations Global SDG Database

### DISCUSSION THREAD

**Discussion by amestsantim** [open-sdg/open-sdg#1906] [General]
I have created a country (Ethiopia) dashboard to use the United Nations Global SDG Database API...

As a starting point, I have managed to configure and build the dashboard ([https://tech-acs.github.io/sdg-ethiopia](https://tech-acs.github.io/sdg-ethiopia)

I struggled to have the input: InputSdmxMl_UnitedNationsApi type working at first but I think I've figured it out...

```yml
inputs:
  - class: InputSdmxMl_UnitedNationsApi
    reference_area: 231
  - class: InputSdmxMeta
    meta_suffix: __GLOBAL
```

The above is the relevant section of the config_data.yml file

My questions are:

- Is it possible to get the national data and not the modeled data from that API? If so, what config keys do I need to use?
- The documentation states that a dimension_query key can be used to set values for dimensions, but I was not able to find any examples of how to use it. Can you please provide an example?
- I would also like to filter by year (basically, I want to display 2000 and above), how do I specify that?
- And last, is there any way of having regional (EAC - East African Community) data show up as one or if not, show up as one row per country?

These are probably silly questions but I just started with OpenSDG so please excuse my ignorance.

Thank you!

**Comment by brockfanning** [comment]
That `InputSdmxMl_UnitedNationsApi` is intended to be a helper class for simple use-cases. So it may not be possible to use that for what you need, but I'll explain more below. FYI, there is documentation available on how to use the UN's API directly here: https://unstats.un.org/sdgs/files/SDMX_SDG_API_MANUAL.pdf

Calls to the UN API should start with this: `https://data.un.org/ws/rest/data/IAEG-SDGs,DF_SDG_GLH/`

Then there should be a "key" (more on that below). And then there can be more after the key. For example, the InputSdmxMl_UnitedNationsApi input puts this at the end: `/ALL/?detail=full&dimensionAtObservation=TIME_PERIOD` (though at the moment I don't know the reason this is added).

To use the API directly, you'll need to understand how the "key" works. The documentation linked above helps with this, but the most relevant parts of that documentation about this are:

> The key has the following structure:
> [FREQ].[REPORTING_TYPE].[SERIES].[REF_AREA].[SEX].[AGE].[URBANISATION].[INCOME_WEALTH_QUANTILE].[EDUCATION_LEV].[OCCUPATION].[CUST_BREAKDOWN].[COMPOSITE_BREAKDOWN].[DISABILITY_STATUS].[ACTIVITY].[PRODUCT]
> • If there is no filter on a dimension, do not provide a value for that dimension.
> • To have the query return observations that match a single value in a particular dimension, insert that value in the placeholder for that dimension. For example, the following key will return all observations with SERIES= SI_POV_DAY1:

> ..SI_POV_DAY1............

So the key needs to be composed of 15 parts, and the 15 parts should be separated by a dot (`.`). In the example above, the results will be filtered to only include data from SI_POV_DAY1 SERIES, but will not be filtered in any other way, since all the other parts are blank.

> Is it possible to get the national data and not the modeled data from that API? If so, what config keys do I need to use?

Yes, to do that you would want the call to the UN API to specify a `REPORTING_TYPE` of `N`. If you are calling the API directly, you would want to have `N` in the 2nd "part" of the key. Eg: `.N.............`

This should give you only national data (as opposed to `G` which gives you global data).

But be aware that at this time there is not as much national data as there is global data, so you may not get many (or any) results with a REPORTING_TYPE of N.

> The documentation states that a dimension_query key can be used to set values for dimensions, but I was not able to find any examples of how to use it. Can you please provide an example?

The dimension_query parameter can be used in InputSdmxMl_UnitedNationsApi as a helpful way to construct the "key" without worrying about all those dots. For example, to specify a REPORTING_STATUS of N you could do:

```
inputs:
  - class: InputSdmxMl_UnitedNationsApi
    reference_area: 231
    dimension_query:
      REPORTING_TYPE: N
```

> I would also like to filter by year (basically, I want to display 2000 and above), how do I specify that?

This is the part of your requirements that I don't think are possible with the InputSdmxMl_UnitedNationsApi class. However it should be possible to get this by constructing the key and using `InputSdmxMl_Structure` instead. The documentation linked above has a section about this. The relevant part is:

> [?startPeriod=yyyy&endPeriod=yyyy] for any optional additional time filtering

For example this might work (I haven't tested, and there may be other parameters I'm forgetting that would be needed):

```
inputs:
  - class: InputSdmxMl_Structure
    source: https://data.un.org/ws/rest/data/IAEG-SDGs,DF_SDG_GLH/...231...........?startPeriod=2000&endPeriod=2022
```

> And last, is there any way of having regional (EAC - East African Community) data show up as one or if not, show up as one row per country?

If I'm understanding correctly, I think you might do this by specifying multiple REF_AREA values. The documentation linked above explains how to query multiple values. The relevent part is:

> To have the query return observations that match any of several values in a particular dimension, separate those values with a plus sign.

Eg: https://data.un.org/ws/rest/data/IAEG-SDGs,DF_SDG_GLH/...231+232...........?startPeriod=2000&endPeriod=2022

**Comment by amestsantim** [comment]
Wow, these are great explanations. 

I suppose InputSdmxMl_UnitedNationsApi is just a specific implementation of InputSdmxMl_Structure and if I want full control of the API fetching, I'd need to use that!

I will try all your suggestions and report back!

Thank you!

**Comment by workualemu** [comment]
I understand this was explained some time ago. I’ve tried using both the InputSdmxMl_UnitedNationsApi and InputSdmxMl_Structure classes as described above, but I keep getting the error: "Unable to parse string NaN."
Do you have any advice on how to resolve this?

`Run python scripts/check_data.py
  python scripts/check_data.py
  python scripts/build_data.py
  shell: /usr/bin/bash -e {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.8.18/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.8.18/x64/lib
00:00:00 - Indicator 15-[2](https://github.com/workualemu/open-sdg-test-data/actions/runs/15743964634/job/44375992888#step:5:2)-1 has a non-numeric value: Unable to parse string "NaN" at position 21
Traceback (most recent call last):
00:00:00 - Indicator 7-1-1 has a non-numeric value: Unable to parse string "NaN" at position 24
  File "scripts/check_data.py", line 8, in <module>
Error:     raise Exception('There were validation errors. See output above.')
Exception: There were validation errors. See output above.
00:00:00 - Indicator 6-5-2 has a non-numeric value: Unable to parse string "NaN" at position 0
00:00:01 - Indicator 6-[3](https://github.com/workualemu/open-sdg-test-data/actions/runs/15743964634/job/44375992888#step:5:3)-1 has a non-numeric value: Unable to parse string "NaN" at position 2
00:00:01 - Indicator 2-5-1 has a non-numeric value: Unable to parse string "NaN" at position [4](https://github.com/workualemu/open-sdg-test-data/actions/runs/15743964634/job/44375992888#step:5:4)6
00:00:01 - Indicator 14-4-1 has a non-numeric value: Unable to parse string "NaN" at position 0
00:00:01 - Indicator 6-4-2 has a non-numeric value: Unable to parse string "NaN" at position 0
00:00:01 - Indicator 6-4-1 has a non-numeric value: Unable to parse string "NaN" at position 0
00:00:01 - Indicator 14-[5](https://github.com/workualemu/open-sdg-test-data/actions/runs/15743964634/job/44375992888#step:5:5)-1 has a non-numeric value: Unable to parse string "NaN" at position 0
00:00:01 - Indicator 15-4-2 has a non-numeric value: Unable to parse string "NaN" at position 2
00:00:01 - Indicator 14-b-1 has a non-numeric value: Unable to parse string "NaN" at position 1
00:00:01 - Indicator 14-[6](https://github.com/workualemu/open-sdg-test-data/actions/runs/15743964634/job/44375992888#step:5:6)-1 has a non-numeric value: Unable to parse string "NaN" at position 0`

**Reply by brockfanning** [reply]
@workualemu Could you paste in the configuration/code of your "inputs"? Or alternatively provide a link to your repository? I don't remember experiencing this issue, but maybe I'll be able to recreate the problem locally with the requested info.

**Comment by workualemu** [comment]
My data repository is at https://github.com/workualemu/open-sdg-test-data
Site repo is at https://github.com/workualemu/open-sdg-test-site
